pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,630 @@
+                                evictions_failed.file_sizes += layer.file_size();
+                                evictions_failed.count += 1;
+                            }
+                            None => {
+                                assert!(cancel.is_cancelled());
+                                return;
+                            }
+                            Some(Err(e)) => {
+                                // we really shouldn't be getting this, precondition failure
+                                error!("failed to evict layer: {:#}", e);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        .instrument(tracing::info_span!("evict_batch", %tenant_id, %timeline_id, batch_size))
+        .await;
+
+        if cancel.is_cancelled() {
+            return Ok(IterationOutcome::Cancelled);
+        }
+    }
+
+    Ok(IterationOutcome::Finished(IterationOutcomeFinished {
+        before: usage_pre,
+        planned: usage_planned,
+        assumed: AssumedUsage {
+            projected_after: usage_assumed,
+            failed: evictions_failed,
+        },
+    }))
+}
+
+async fn extend_lru_candidates(
+    tenant_min_resident_size: Option<u64>,
+    tenant_id: TenantId,
+    lru_candidates: &mut Vec<(Arc<Timeline>, LocalLayerInfoForDiskUsageEviction)>,
+    scratch: &mut Vec<(Arc<Timeline>, LocalLayerInfoForDiskUsageEviction)>,
+    cancel: &CancellationToken,
+) -> ControlFlow<()> {
+    let tenant = match tenant::mgr::get_tenant(tenant_id, true).await {
+        Ok(tenant) => tenant,
+        // this can happen if tenant has lifecycle transition after we fetched it above
+        Err(e) => {
+            debug!("failed to get tenant {tenant_id}: {e:#}");
+            return ControlFlow::Continue(());
+        }
+    };
+
+    if cancel.is_cancelled() {
+        return ControlFlow::Break(());
+    }
+
+    let timelines = tenant.list_timelines();
+    for tl in timelines {
+        if !tl.is_active() {
+            continue;
+        }
+        scratch.extend(
+            tl.get_local_layers_for_disk_usage_eviction()
+                .into_iter()
+                .map(|layer_infos| (tl.clone(), layer_infos)),
+        );
+        if cancel.is_cancelled() {
+            return ControlFlow::Break(());
+        }
+    }
+
+    // sort layers by last activity
+    scratch.sort_unstable_by_key(|(_, layer_info)| layer_info.last_activity_ts);
+
+    let min_resident_size = tenant_min_resident_size.unwrap_or(0); // XXX allow per-tenant override, including "unlimited"
+    let mut post_evict_resident_size: u64 =
+        scratch.iter().map(|(_, layer)| layer.file_size()).sum();
+
+    for (tl, layer) in scratch.drain(..) {
+        if cancel.is_cancelled() {
+            return ControlFlow::Break(());
+        }
+        if post_evict_resident_size <= min_resident_size {
+            continue;
```

Comment:
This looks like it should be a `break`; will be changing it.

Reply:
Changed in [0fd78c3](https://github.com/neondatabase/neon/pull/3809/commits/0fd78c3d421611ab502a0b3f109f74ba1b1dda48)